src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs

```diff
@@ -214,6 +220,24 @@ private ValueTask<HttpConnection> GetConnectionAsync(HttpRequestMessage request,
                 new HttpConnectionWithFinalizer(this, stream, transportContext); // finalizer needed to signal the pool when a connection is dropped
         }
 
+        private async ValueTask<Stream> EstablishProxyTunnel(CancellationToken cancellationToken)
+        {
+            // Send a CONNECT request to the proxy server to establish a tunnel.
+            HttpRequestMessage tunnelRequest = new HttpRequestMessage(new HttpMethod("CONNECT"), _proxyUri);
+            tunnelRequest.Headers.Host = $"{_host}:{_port}";    // This specifies destination host/port to connect to
```

Comment:
And can we cache the whole `tunnelRequest` object on this pool so that there's just one per pool?

Reply:
The whole object?  You mean reuse the same request object for every CONNECT request?  That seems a little dangerous.

Reply:
> The whole object?

Yes.

> That seems a little dangerous

Depends where we land with whether and how we end up modifying the HttpRequestMessage, and whether such modifications happen in handlers below this point.

Reply:
Until then though, I'd prefer to just create a new request.  Given the overhead of establishing the proxy tunnel, this doesn't seem like it matters in practice.

Reply:
Ok